

# <code style="background:yellow;color:black">Summary of today's session:</code> 

1. What is pandas and why do we use pandas?<br>

2. Basic data structures of pandas -- **<code style="background:yellow;color:black">Series and DataFrame.</code>**<br>

3. **<code style="background:yellow;color:black">pd.read_csv() function</code>** to load or import the csv dataset into pandas.<br>

4. **<code style="background:yellow;color:black">Inspecting the DataFrame:</code>**
    * The **<code style="background:yellow;color:black">head() method</code>** used to access first few rows of a DataFrame or Series.
    * Just writing the name of dataframe and executing it will give us the entire dataset as output on screen.
    * **<code style="background:yellow;color:black">type() function</code> is a python function used to know the type of object, whether its a dataframe or series.**
    * The **<code style="background:yellow;color:black">tail() method</code>** is used to access last few rows of a DataFrame or Series.
    * If we want to find the dimensions or shape of the dataframe, we can use the **<code style="background:yellow;color:black">shape attribute.</code>**
    * **<code style="background:yellow;color:black">info() method</code>** tells us how many rows and columns are there in our dataframe, for each column how many null and not null values are there, datatype of all those values and the memory usage of the dataframe.
    * The **<code style="background:yellow;color:black">columns attribute</code>** is used to access and view the column labels (column names) of a DataFrame. Indexing can be used on top of it to fetch the exact column name.
    * If we want to access a particular column in a pandas DataFrame, we can do it by using square brackets [ ] and writing code as **<code style="background:yellow;color:black">df["col_name"].</code>** Similarly, if we want to access multiple columns, we will provide a list of column names inside **<code style="background:yellow;color:black">df["col_1", col_2"].</code>**<br>

5. **<code style="background:yellow;color:black">Creating a DataFrame from scratch: pd.DataFrame()</code>** is a function used to create a dataset from scratch. There are two approaches called row oriented and column oriented approach to create a new dataframe from scratch.<br>

6. **<code style="background:yellow;color:black">Basic operations on a column:</code>**

    * Using **<code style="background:yellow;color:black">unique() method</code>** we will get all the unique values in a column. Kind of like "distinct" in sql.
    * **<code style="background:yellow;color:black">nunique() method</code>** gives us a count or total number of unique values in a column.
    * **<code style="background:yellow;color:black">value_counts() method</code>** is very very frequently used method. It returns for every unique value, how many times its occuring in that column.
    * Renaming a column- **<code style="background:yellow;color:black">rename() method</code>** is used to rename column names. To rename a column, we pass a dictionary in the rename() method, where, dict key is the existing column name and dict value is the new name that we want to update. Now in one way it does not modify the original dataframe, it just creates a copy of it and returns us that. In another way, by passing an argument inplace = True, we can modify the original dataframe itself. 
    * Deleting a column- There is a method called **<code style="background:yellow;color:black">drop() method</code>** which deletes columns. One syntax is df.drop("col_name", axis = 1). Here inplace argument can be supplied as per requirement. Another syntax is new_df.drop(columns = ["col_1", "col_2", "col_3"]). Here axis argument is not required and multiple columns can be deleted in one go. Again inplace argument can be supplied as per requirement.
    * **<code style="background:yellow;color:black">Creating a column-</code>** We can create a new column from existing columns by performing some arithmetic operation or we can create a new column from scratch.<br>

7. **<code style="background:yellow;color:black">Dataframe indexing-</code>** There are two types of indexes called <code style="background:yellow;color:black">implicit and expicit index.</code> There are various indexing options and ways available to index columns and rows including loc and iloc indexers. In other words with these ways we can acccess rows and columns of our dataset.<br>

8. Setting the index to some other value by using **<code style="background:yellow;color:black">df.set_index("index_name") method.</code>**<br>

9. Resetting the index back to original can be done using **<code style="background:yellow;color:black">df.reset_index() method.</code>**  

10. **<code style="background:yellow;color:black">Difference between pandas method and function.</code>** 

<hr style="border: 1px solid gray;">

# <code style="background:yellow;color:black">PANDAS:</code>

### <code style="background:yellow;color:black">What is pandas?</code> 

**<code style="background:yellow;color:black">Pandas is an open-source data manipulation and analysis library</code> for the Python programming language. It provides easy-to-use data structures and data analysis tools for working with structured data, such as tabular data, time series, and more. Pandas is a popular choice among data scientists, analysts, and engineers for a variety of data-related tasks.** 

### <code style="background:yellow;color:black">Why do we use pandas?</code> 

1. **<u>Data Import and Export:</u>** Pandas makes it easy to import data from various file formats, such as CSV, Excel, SQL databases, JSON, and more. You can also export data to these formats.

2. **<u>Data Cleaning and Preprocessing:</u>** Pandas offers a wide range of functions for data cleaning, transformation, and preprocessing. This includes handling missing data, removing duplicates, converting data types, and more.

3. **<u>Data Exploration and Analysis:</u>** You can perform various data analysis tasks with Pandas, such as filtering, sorting, grouping, aggregating, and computing descriptive statistics. It allows you to gain insights into your data quickly.

4. **<u>Time Series Analysis:</u>** Pandas has built-in support for time series data, making it well-suited for working with temporal data and conducting time-based analysis.

5. **<u>Data Visualization:</u>** While Pandas itself is not a data visualization library, it can be easily integrated with libraries like Matplotlib and Seaborn for data visualization.

6. **<u>Integration with Other Libraries:</u>** Pandas can seamlessly integrate with other Python libraries commonly used in data science, such as NumPy for numerical operations and scikit-learn for machine learning.

7. **<u>Performance and Efficiency:</u>** Pandas is designed for high performance and memory efficiency, making it suitable for working with large datasets.

### <code style="background:yellow;color:black">Series and Dataframe:</code>

**<code style="background:yellow;color:black">Pandas introduces two primary data structures, namely, Series and DataFrame.</code>**

* <u>**Series:**</u> A one-dimensional array-like object that can hold data of any data type. It's similar to a column in a spreadsheet or a single feature in a dataset.

* <u>**DataFrame:**</u> A two-dimensional tabular data structure that resembles a table with rows and columns. It's used to store and manipulate structured data efficiently.

Whenever we are working with a data which is of different formats, like for example in excel, we have some integer datatype columns some string datatype column and some can be float, so whenever we are working with data where different columns have different datatypes, **that dataset WE CALL IT PANDAS DATAFRAME**. A Dataframe is nothing but a table in form of rows and columns just like sql tables or excel spreadsheets. 

In [6]:
import numpy as np
import pandas as pd

<hr style="border: 1px solid gray;">

# <code style="background:yellow;color:black">Loading or Importing a DataFrame:</code>

**Whenever we have to load a dataset, we execute the following code. We are loading the dataset called "mckinsey" in a variable called "df", which is a dataframe. You can call it anything, its just a variable but its a dataframe.**

**The <code style="background:yellow;color:black">pd.read_csv() function</code> is a widely used function in the pandas library, which allows us to read data from CSV (Comma-Separated Values) files and create a DataFrame, which is a fundamental data structure in pandas for handling and analyzing tabular data.**

**<u>Following are the functions to read or load data in pandas dataframe:**</u>

1. **pd.read_excel('filename.xlsx'):** Reads an Excel file into a DataFrame.
2. **pd.read_json('filename.json'):** Reads a JSON file into a DataFrame.
3. **pd.read_sql():** Reads a SQL table or query into a DataFrame using SQLAlchemy.
4. **pd.read_html('url_or_html_file'):** Reads HTML tables into a list of DataFrame objects.
5. **pd.read_feather('filename.feather'):** Reads a feather-format file into a DataFrame. Feather is a binary columnar data format.
6. **pd.read_parquet('filename.parquet'):** Reads a Parquet file into a DataFrame. Parquet is a columnar storage file format.
7. **pd.read_clipboard():** Reads data from the clipboard. This can be useful when you've copied a table from a website or another source.

In [7]:
df = pd.read_csv("mckinsey.csv")

<hr style="border: 1px solid gray;">

# <code style="background:yellow;color:black">Inspecting a DataFrame:</code>

### <code style="background:yellow;color:black">1. head() method:</code>

**The <code style="background:yellow;color:black">head() method</code> is used to access first few rows of a DataFrame or Series**. 

* It is a convenient way to quickly inspect the data in a tabular format and get a sense of what the data looks like without displaying the entire DataFrame. 
* By default, head() displays the first 5 rows.
* Here below we can see that our data is in form of rows and columns. There are 5 rows and 6 columns. 
* **The first column is called "index" and its not considered as a column**. 
* If we write simply DataFrame name "df" it will give us the entire dataframe means the entire dataset. We can see in the code cell below that there are total 1704 rows and 6 columns in our entire dataset. Generally, it displays in the format like first 5 rows, dot dot dot and last 5 rows.
* **If we want to access some specific amount of rows, we can provide the number of rows as a parameter in the head() method**. Like in below code we are specifying 10, means it will return 10 rows. **<code style="background:yellow;color:black">This is very similar to "limit" clause in sql.</code>**

In [8]:
df.head()

country  year  population continent  life_exp     gdp_cap
0  Afghanistan  1952     8425333      Asia    28.801  779.445314
1  Afghanistan  1957     9240934      Asia    30.332  820.853030
2  Afghanistan  1962    10267083      Asia    31.997  853.100710
3  Afghanistan  1967    11537966      Asia    34.020  836.197138
4  Afghanistan  1972    13079460      Asia    36.088  739.981106

In [9]:
df

country  year  population continent  life_exp     gdp_cap
0     Afghanistan  1952     8425333      Asia    28.801  779.445314
1     Afghanistan  1957     9240934      Asia    30.332  820.853030
2     Afghanistan  1962    10267083      Asia    31.997  853.100710
3     Afghanistan  1967    11537966      Asia    34.020  836.197138
4     Afghanistan  1972    13079460      Asia    36.088  739.981106
...           ...   ...         ...       ...       ...         ...
1699     Zimbabwe  1987     9216418    Africa    62.351  706.157306
1700     Zimbabwe  1992    10704340    Africa    60.377  693.420786
1701     Zimbabwe  1997    11404948    Africa    46.809  792.449960
1702     Zimbabwe  2002    11926563    Africa    39.989  672.038623
1703     Zimbabwe  2007    12311143    Africa    43.487  469.709298

[1704 rows x 6 columns]

In [10]:
df.head(10)

country  year  population continent  life_exp     gdp_cap
0  Afghanistan  1952     8425333      Asia    28.801  779.445314
1  Afghanistan  1957     9240934      Asia    30.332  820.853030
2  Afghanistan  1962    10267083      Asia    31.997  853.100710
3  Afghanistan  1967    11537966      Asia    34.020  836.197138
4  Afghanistan  1972    13079460      Asia    36.088  739.981106
5  Afghanistan  1977    14880372      Asia    38.438  786.113360
6  Afghanistan  1982    12881816      Asia    39.854  978.011439
7  Afghanistan  1987    13867957      Asia    40.822  852.395945
8  Afghanistan  1992    16317921      Asia    41.674  649.341395
9  Afghanistan  1997    22227415      Asia    41.763  635.341351

<hr style="border: 1px solid gray;">

### <code style="background:yellow;color:black">2. type() function:</code>

**The <code style="background:yellow;color:black">type() function is a built-in Python function</code> that is used to determine the type of an object. It can be used to check the data type of various objects, including pandas DataFrames and Series.**

* **It is not specific to pandas but is a general-purpose function in Python.** 
* **If we do type(df), it will tell us the datatype of df**. 
* So in the output of the code below, it will return pandas.core.frame.DataFrame, which means the type of datatype of df is a pandas dataframe.

In [11]:
type(df)

pandas.core.frame.DataFrame

<hr style="border: 1px solid gray;">

### <code style="background:yellow;color:black">3. tail() method:</code>

**The <code style="background:yellow;color:black">tail() method</code> is used to view the last few rows of a DataFrame or Series.** 

* **It is similar to the head() method, which displays the first few rows, but tail() displays the last rows of the data.** 
* By default, tail() shows the last 5 rows, but we can specify the number of rows we want to see by passing an argument to this tail() method, just like we did in head() method.

In [12]:
df.tail()

country  year  population continent  life_exp     gdp_cap
1699  Zimbabwe  1987     9216418    Africa    62.351  706.157306
1700  Zimbabwe  1992    10704340    Africa    60.377  693.420786
1701  Zimbabwe  1997    11404948    Africa    46.809  792.449960
1702  Zimbabwe  2002    11926563    Africa    39.989  672.038623
1703  Zimbabwe  2007    12311143    Africa    43.487  469.709298

In [13]:
df.tail(7)

country  year  population continent  life_exp     gdp_cap
1697  Zimbabwe  1977     6642107    Africa    57.674  685.587682
1698  Zimbabwe  1982     7636524    Africa    60.363  788.855041
1699  Zimbabwe  1987     9216418    Africa    62.351  706.157306
1700  Zimbabwe  1992    10704340    Africa    60.377  693.420786
1701  Zimbabwe  1997    11404948    Africa    46.809  792.449960
1702  Zimbabwe  2002    11926563    Africa    39.989  672.038623
1703  Zimbabwe  2007    12311143    Africa    43.487  469.709298

<hr style="border: 1px solid gray;">

### <code style="background:yellow;color:black">4. shape attribute:</code>

**If we want to find the dimensions or shape of the dataframe, we can use the <code style="background:yellow;color:black">.shape attribute.</code>**

* This was also there in numpy. Here in this example below, we can see that shape of our df is (1704, 6), means its having 1704 rows and 6 columns. 
* **Output of .shape is always a tuple. (1704, 6) is a tuple.** 
* So datatype or type(shape) is a tuple. 

In [14]:
df.shape

(1704, 6)

<hr style="border: 1px solid gray;">

### <code style="background:yellow;color:black">5. info() method:</code>

**<code style="background:yellow;color:black">info() method</code> is a pre-defined method in pandas and it stands for information, which gives us lots of information about our dataframe.** 

* **It tells us how many rows and columns are there, for each column what are the null and not null values, datatype of all those values and the memory usage of the dataframe**. 
* Now null means missing value. So, in below case, we can see it returns every column's info and that every column has 1704 non-null values. 
* Now because there are 1704 rows or entries, this means there is some value in every row of every column and that there is no null value present in it. 

**<code style="background:yellow;color:black">Also note that</code> whenever we are working with pandas, <code style="background:yellow;color:black">string datatype will be referred as object.</code>**   

In [15]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1704 entries, 0 to 1703
Data columns (total 6 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   country     1704 non-null   object 
 1   year        1704 non-null   int64  
 2   population  1704 non-null   int64  
 3   continent   1704 non-null   object 
 4   life_exp    1704 non-null   float64
 5   gdp_cap     1704 non-null   float64
dtypes: float64(2), int64(2), object(2)
memory usage: 80.0+ KB


<hr style="border: 1px solid gray;">

### <code style="background:yellow;color:black">6. columns attribute:</code>

**The <code style="background:yellow;color:black">columns attribute</code> is used to access and view the column labels (column names) of a DataFrame. It returns an Index object containing the names of all the columns in the DataFrame.** 

* **We can use this attribute to inspect the column names, change column names, or access specific columns by their names.**
* Below code shows how we can access column names. 0th index value is 1st column called country.
* Similarly, 4th index value is 5th column called life_exp.

In [16]:
df.columns

Index(['country', 'year', 'population', 'continent', 'life_exp', 'gdp_cap'], dtype='object')

In [17]:
df.columns[0]

'country'

In [18]:
df.columns[4]

'life_exp'

<hr style="border: 1px solid gray;">

### <code style="background:yellow;color:black">7. square brackets []:</code>

**<code style="background:yellow;color:black">If we want to access a particular column in a pandas DataFrame, we can do it by using square brackets []</code>.** 

* **We can implement code like this as shown below**. Here it returns all 1704 rows of "country" column. 
* **Now when we access a particular column in the below syntax, it returns something called "PANDAS SERIES", which is nothing but a column.** 
* Series is same as numpy array or python lists. 
* If we do type(df["country"]), we will get -- pandas.core.series.Series. So its a series.

In [19]:
df["country"]

0       Afghanistan
1       Afghanistan
2       Afghanistan
3       Afghanistan
4       Afghanistan
           ...     
1699       Zimbabwe
1700       Zimbabwe
1701       Zimbabwe
1702       Zimbabwe
1703       Zimbabwe
Name: country, Length: 1704, dtype: object

In [20]:
type(df["country"])

pandas.core.series.Series

* **Below is other way to get all rows of a particular column or in other words get pandas series.**
* Works well, but we dont use this in industry, because if our column name has a whitespace we connot write it there, it will not work out.

In [21]:
df.country

0       Afghanistan
1       Afghanistan
2       Afghanistan
3       Afghanistan
4       Afghanistan
           ...     
1699       Zimbabwe
1700       Zimbabwe
1701       Zimbabwe
1702       Zimbabwe
1703       Zimbabwe
Name: country, Length: 1704, dtype: object

* Multi-indexing in numpy refers to the fact that we can access multiple indices. In that, we provide a list of indices to access. 
* **Similarly in pandas, <code style="background:yellow;color:black">if we want to access multiple columns,</code> we will have to provide a list of column names inside df[ ], as implemented below**. 
* A new dataframe will be created from the original dataframe with only 2 columns which we wanted. Sequence of column names provided in list as a parameter is important because in that sequence only columns will appear in output.
* This new dataframe will be stored in new variable called "df2".

In [22]:
sample_df = df[["year", "country"]]

sample_df.head()

year      country
0  1952  Afghanistan
1  1957  Afghanistan
2  1962  Afghanistan
3  1967  Afghanistan
4  1972  Afghanistan

<hr style="border: 1px solid gray;">

# <code style="background:yellow;color:black">Creating a DataFrame:</code>

### <code style="background:yellow;color:black">pd.DataFrame():</code>

* **<code style="background:yellow;color:black">pd.DataFrame()</code> is a function used to create a dataset from scratch**. 
* **So, far we have loaded the pre-existing dataset. But we want to create a table from scratch.** 
* What if we want to create a dataframe from scratch, how do we do that? 
* Lets say we want to create the below dataset: There are 3 rows and 6 columns, and we want to create this dataset from scratch.

|  |country     |year |population |continent |life_exp |gdp_cap    |
|:-|:-----------|:----|:----------|:---------|:--------|:----------|
|0 |Afghanistan |1952 |8425333    |Asia      |28.801   |779.445314 | 
|1 |Afghanistan |1957 |9240934    |Asia      |30.332   |820.853030 |
|2 |Afghanistan |1962 |102267083  |Asia      |31.997   |853.100710 |

**<u>There are two approaches to create a DataFrame from pd.DataFrame() function:**</u>

**<code style="background:yellow;color:black">1. Row-oriented approach:</code>**

* **Its called row-oriented approach because we are providing one row at a time.** 
* **Within this pd.DataFrame() function, we have to provide 2 parameters.** 
* The 1st parameter is a 2D list. As we can see in below example, its a nested list where every row is represented with one list. There are 3 rows as it was required, we have created that. 
* Now for the 2nd parameter, first we put a comma and in 2nd parameter we give column names as a list. Column names are given in comma seperated sequence. 
* So if we run sample_df, we can see the created dataset.

In [23]:
df1 = pd.DataFrame(
              [
                  ['Afghanistan', 1952, 8425333, 'Asia', 28.801, 779.445314 ],
                  ['Afghanistan', 1957, 9240934, 'Asia', 30.332, 820.853030 ],
                  ['Afghanistan', 1962, 102267083, 'Asia', 31.997, 853.100710 ]
              ],

             columns = ['country','year','population','continent','life_exp','gdp_cap']
    )

df1

country  year  population continent  life_exp     gdp_cap
0  Afghanistan  1952     8425333      Asia    28.801  779.445314
1  Afghanistan  1957     9240934      Asia    30.332  820.853030
2  Afghanistan  1962   102267083      Asia    31.997  853.100710

**<code style="background:yellow;color:black">2. Column-oriented approach:</code>**

* **Here we are using same pd.Dataframe() function to create our own dataset, but we are providing dictionary as a parameter**. * We have covered our dictionary session, we know dictionary is a collection of key value pairs. 
* **Here, column name is the dict key and list of column values or rows is dict value**. 
* There is only one parameter to be provided in column oriented approach.

In [24]:
df2 = pd.DataFrame(
              {
                  'country' : ['Afghanistan', 'Afghanistan'],
                   'year' : [1952,1957],
                  'population' : [842533, 9240934],
                   'continent' : ['Asia', 'Asia'],
                  'life_exp' : [28.801, 30.332],
                   'gdp_cap' : [779.445314, 820.853030]
               }
    )

df2

country  year  population continent  life_exp     gdp_cap
0  Afghanistan  1952      842533      Asia    28.801  779.445314
1  Afghanistan  1957     9240934      Asia    30.332  820.853030

<hr style="border: 1px solid gray;">

# <code style="background:yellow;color:black">Basic operations on column:</code>

### <code style="background:yellow;color:black">1. unique() method:</code>

* **Using <code style="background:yellow;color:black">unique() method</code>, we will get all the unique values in a column.**
* The unique() method in pandas is used to get the unique values in a pandas Series. 
* It returns an array of unique elements in the order in which they appear in the original Series 
* **No duplicates are returned. Kind of like <code style="background:yellow;color:black">"distinct"</code> in sql.**
* The unique() method is useful when we want to identify the distinct values within a pandas Series or a specific column of a DataFrame.

In [25]:
unique_country = df["country"].unique() 

unique_country

array(['Afghanistan', 'Albania', 'Algeria', 'Angola', 'Argentina',
       'Australia', 'Austria', 'Bahrain', 'Bangladesh', 'Belgium',
       'Benin', 'Bolivia', 'Bosnia and Herzegovina', 'Botswana', 'Brazil',
       'Bulgaria', 'Burkina Faso', 'Burundi', 'Cambodia', 'Cameroon',
       'Canada', 'Central African Republic', 'Chad', 'Chile', 'China',
       'Colombia', 'Comoros', 'Congo, Dem. Rep.', 'Congo, Rep.',
       'Costa Rica', "Cote d'Ivoire", 'Croatia', 'Cuba', 'Czech Republic',
       'Denmark', 'Djibouti', 'Dominican Republic', 'Ecuador', 'Egypt',
       'El Salvador', 'Equatorial Guinea', 'Eritrea', 'Ethiopia',
       'Finland', 'France', 'Gabon', 'Gambia', 'Germany', 'Ghana',
       'Greece', 'Guatemala', 'Guinea', 'Guinea-Bissau', 'Haiti',
       'Honduras', 'Hong Kong, China', 'Hungary', 'Iceland', 'India',
       'Indonesia', 'Iran', 'Iraq', 'Ireland', 'Israel', 'Italy',
       'Jamaica', 'Japan', 'Jordan', 'Kenya', 'Korea, Dem. Rep.',
       'Korea, Rep.', 'Kuwait', 'Leba

### <code style="background:yellow;color:black">2. nunique() method:</code>

* **<code style="background:yellow;color:black">nunique() method</code> gives us a count or total number of unique values in a column.**
* Its very similar method as unique() method.
* The nunique() method in pandas is used to count the number of unique values in a Series or DataFrame column. It returns the count of distinct elements present in the specified column.
* The nunique() method is useful when we want to quickly determine the number of distinct elements in a column, providing a summary statistic for the diversity of values in the data.

In [26]:
count_unique = df["country"].nunique()

count_unique

142

In [27]:
df.nunique()

country        142
year            12
population    1704
continent        5
life_exp      1626
gdp_cap       1704
dtype: int64

In [28]:
df.unique()

AttributeError: 'DataFrame' object has no attribute 'unique'

### <code style="background:yellow;color:black">3. value_counts() method:</code>

* **<code style="background:yellow;color:black">value_counts() method</code> is very very frequently used function.** 
* **It returns for every unique value, how many times its occuring in that column.** 
* The value_counts() method in pandas is used to compute a histogram of a Series, providing counts of unique values in the Series. It returns a new Series with the unique values as the index and the corresponding counts as the values.
* In this example output, we can see that every country's name is appearing 12 times. 
* **<code style="background:yellow;color:black">This is very similar to "group by" in sql.</code> It is like group by country name and count how many entries of each country is there.**
* **value_counts() will always return the data in decreasing order of sorting**. We can also do that sorting in ascending order by providing a parameter to the function.
* The value_counts() method is particularly useful when you want to understand the distribution of values within a categorical variable in your data. It provides a quick summary of how many times each unique value appears in the Series or DataFrame column.

In [29]:
df["country"].value_counts()

Afghanistan          12
Pakistan             12
New Zealand          12
Nicaragua            12
Niger                12
                     ..
Eritrea              12
Equatorial Guinea    12
El Salvador          12
Egypt                12
Zimbabwe             12
Name: country, Length: 142, dtype: int64

In [30]:
df["continent"].value_counts()

Africa      624
Asia        396
Europe      360
Americas    300
Oceania      24
Name: continent, dtype: int64

### <code style="background:yellow;color:black">4. Renaming columns:</code>

**<code style="background:yellow;color:black">rename() method</code> is used to rename column names.** 

* **To rename a column name, we pass a dictionary in the rename() method, where dict key is the existing column name and dict value is the new name that we want to update.** 
* Here below, we implemented that code, but column names did not get updated as we can see in output, because by default axis is 0, and axis 0 refers to vertical axis. So when we are not giving any axis, it will take default axis as 0; it is looking for column names over the indices where no column names are present, basically its searching vertically the column names to update. 
* So in the next code cell below to below this code, we supply axis = 1. Now, we get the updated column names in the output.

In [31]:
df.head()

country  year  population continent  life_exp     gdp_cap
0  Afghanistan  1952     8425333      Asia    28.801  779.445314
1  Afghanistan  1957     9240934      Asia    30.332  820.853030
2  Afghanistan  1962    10267083      Asia    31.997  853.100710
3  Afghanistan  1967    11537966      Asia    34.020  836.197138
4  Afghanistan  1972    13079460      Asia    36.088  739.981106

In [32]:
df.rename(
    {
        "population" : "Population",
         "country" : "Country"}
)

country  year  population continent  life_exp     gdp_cap
0     Afghanistan  1952     8425333      Asia    28.801  779.445314
1     Afghanistan  1957     9240934      Asia    30.332  820.853030
2     Afghanistan  1962    10267083      Asia    31.997  853.100710
3     Afghanistan  1967    11537966      Asia    34.020  836.197138
4     Afghanistan  1972    13079460      Asia    36.088  739.981106
...           ...   ...         ...       ...       ...         ...
1699     Zimbabwe  1987     9216418    Africa    62.351  706.157306
1700     Zimbabwe  1992    10704340    Africa    60.377  693.420786
1701     Zimbabwe  1997    11404948    Africa    46.809  792.449960
1702     Zimbabwe  2002    11926563    Africa    39.989  672.038623
1703     Zimbabwe  2007    12311143    Africa    43.487  469.709298

[1704 rows x 6 columns]

In [33]:
df.rename(
    {
        "population" : "Population",
         "country" : "Country"},

    axis = 1
)

Country  year  Population continent  life_exp     gdp_cap
0     Afghanistan  1952     8425333      Asia    28.801  779.445314
1     Afghanistan  1957     9240934      Asia    30.332  820.853030
2     Afghanistan  1962    10267083      Asia    31.997  853.100710
3     Afghanistan  1967    11537966      Asia    34.020  836.197138
4     Afghanistan  1972    13079460      Asia    36.088  739.981106
...           ...   ...         ...       ...       ...         ...
1699     Zimbabwe  1987     9216418    Africa    62.351  706.157306
1700     Zimbabwe  1992    10704340    Africa    60.377  693.420786
1701     Zimbabwe  1997    11404948    Africa    46.809  792.449960
1702     Zimbabwe  2002    11926563    Africa    39.989  672.038623
1703     Zimbabwe  2007    12311143    Africa    43.487  469.709298

[1704 rows x 6 columns]

* **The most important thing to note is when we check our dataset again by doing df.head(), we see that our column names are not changed.** 
* When we write the code as above in rename() method, **<code style="background:yellow;color:black">it does not modify the original dataframe.</code>** It creates a copy of it and returns us that. We havent modified original df itself. 
* So above syntax is creating a copy of original dataframe which is containing the modified column names. Thats why when below we do df.head(), the original dataframe is not changed.
* **One way by which we can collect or store the copy of modified data is to collect it in a new variable called new_df as done below.**
* But again we see original dataframe is not updated by checking df.head().
* **<code style="background:yellow;color:black">We can modify the changes in the original dataframe itself by providing a parameter called "inplace = True".</code>** This means we are telling pandas that dont create a copy, modify the original dataframe itself. 
* By default, inplace is false. Below we are implementing inplace = True. This doesnt return anything because its modifying the original dataframe. No copies being returned, unlike in the above code we saw that copy was returned as an output. 
* Now, if we check using df.head(), we see original dataframe is changed.

In [34]:
df.head()

country  year  population continent  life_exp     gdp_cap
0  Afghanistan  1952     8425333      Asia    28.801  779.445314
1  Afghanistan  1957     9240934      Asia    30.332  820.853030
2  Afghanistan  1962    10267083      Asia    31.997  853.100710
3  Afghanistan  1967    11537966      Asia    34.020  836.197138
4  Afghanistan  1972    13079460      Asia    36.088  739.981106

In [35]:
new_df = df.rename(
    {"population" : "Population",
     "country" : "Country"},

    axis = 1
)

In [36]:
new_df

Country  year  Population continent  life_exp     gdp_cap
0     Afghanistan  1952     8425333      Asia    28.801  779.445314
1     Afghanistan  1957     9240934      Asia    30.332  820.853030
2     Afghanistan  1962    10267083      Asia    31.997  853.100710
3     Afghanistan  1967    11537966      Asia    34.020  836.197138
4     Afghanistan  1972    13079460      Asia    36.088  739.981106
...           ...   ...         ...       ...       ...         ...
1699     Zimbabwe  1987     9216418    Africa    62.351  706.157306
1700     Zimbabwe  1992    10704340    Africa    60.377  693.420786
1701     Zimbabwe  1997    11404948    Africa    46.809  792.449960
1702     Zimbabwe  2002    11926563    Africa    39.989  672.038623
1703     Zimbabwe  2007    12311143    Africa    43.487  469.709298

[1704 rows x 6 columns]

In [37]:
df.head()

country  year  population continent  life_exp     gdp_cap
0  Afghanistan  1952     8425333      Asia    28.801  779.445314
1  Afghanistan  1957     9240934      Asia    30.332  820.853030
2  Afghanistan  1962    10267083      Asia    31.997  853.100710
3  Afghanistan  1967    11537966      Asia    34.020  836.197138
4  Afghanistan  1972    13079460      Asia    36.088  739.981106

In [38]:
df.rename(
    {
        "population" : "Population",
         "country" : "Country"
    },

    axis = 1,
    inplace = True
)

In [39]:
df.head()

Country  year  Population continent  life_exp     gdp_cap
0  Afghanistan  1952     8425333      Asia    28.801  779.445314
1  Afghanistan  1957     9240934      Asia    30.332  820.853030
2  Afghanistan  1962    10267083      Asia    31.997  853.100710
3  Afghanistan  1967    11537966      Asia    34.020  836.197138
4  Afghanistan  1972    13079460      Asia    36.088  739.981106

### <code style="background:yellow;color:black">5. Deleting columns:</code>

**There is a method called <code style="background:yellow;color:black">drop() method</code> which is used for deleting columns.**

**There are two ways of deleting a column:**

**<code style="background:yellow;color:black">First way</code> is like the one we are implementing below.** 
* It returns us a copy of df with deleted column. Thats why when in next line we check original data using df.head(), we see original data hasnt been updated or continent column hasnt been deleted. The change is not permanent.
* So if we want to modify the original dataframe, we have to provide the "inplace = True" parameter in the drop() method, which we have done in the next code cell.
* This will not return anything, because it has updated the original dataframe. Thats why when we check original dataframe using df.head(), we can see original df changed i.e. continent column deleted.
* **Please note that deleted original data cant be retrieved, so make a copy of it if possible.**

In [40]:
df.drop("continent", axis = 1)

Country  year  Population  life_exp     gdp_cap
0     Afghanistan  1952     8425333    28.801  779.445314
1     Afghanistan  1957     9240934    30.332  820.853030
2     Afghanistan  1962    10267083    31.997  853.100710
3     Afghanistan  1967    11537966    34.020  836.197138
4     Afghanistan  1972    13079460    36.088  739.981106
...           ...   ...         ...       ...         ...
1699     Zimbabwe  1987     9216418    62.351  706.157306
1700     Zimbabwe  1992    10704340    60.377  693.420786
1701     Zimbabwe  1997    11404948    46.809  792.449960
1702     Zimbabwe  2002    11926563    39.989  672.038623
1703     Zimbabwe  2007    12311143    43.487  469.709298

[1704 rows x 5 columns]

In [41]:
df.head()

Country  year  Population continent  life_exp     gdp_cap
0  Afghanistan  1952     8425333      Asia    28.801  779.445314
1  Afghanistan  1957     9240934      Asia    30.332  820.853030
2  Afghanistan  1962    10267083      Asia    31.997  853.100710
3  Afghanistan  1967    11537966      Asia    34.020  836.197138
4  Afghanistan  1972    13079460      Asia    36.088  739.981106

In [42]:
df.drop("continent", axis = 1, inplace = True)

df.head()

Country  year  Population  life_exp     gdp_cap
0  Afghanistan  1952     8425333    28.801  779.445314
1  Afghanistan  1957     9240934    30.332  820.853030
2  Afghanistan  1962    10267083    31.997  853.100710
3  Afghanistan  1967    11537966    34.020  836.197138
4  Afghanistan  1972    13079460    36.088  739.981106

**<code style="background:yellow;color:black">Second way</code> to delete columns is like the one we are implementing below.**
* Method is same drop() method, but syntax is slight different. 
* **Here, note there is no need to provide axis, because we are specifically saying columns = ["continent"].**
* Again, code below will return a copy of modified df. This approach is used when we dont want to use axis.
* In the next code below, the df is changed permanently, the continent column is permanenetly deleted from the original dataframe. So, wherever we want to make change permanent, we have to write "inplace = True". We check original dataframe using df.head(), and we can see original df changed i.e. continent column deleted.
* **Lets say, we want to delete multiple columns, we do that by providing multiple column names like the code implemented below.** 
* So in the next code cell, we are deleting "life_exp and year" columns permanently. We check original dataframe using df.head(), and we can see original df changed i.e. "life_exp and year" columns have been deleted.

In [43]:
new_df.head() # This df we created during renaming column names

Country  year  Population continent  life_exp     gdp_cap
0  Afghanistan  1952     8425333      Asia    28.801  779.445314
1  Afghanistan  1957     9240934      Asia    30.332  820.853030
2  Afghanistan  1962    10267083      Asia    31.997  853.100710
3  Afghanistan  1967    11537966      Asia    34.020  836.197138
4  Afghanistan  1972    13079460      Asia    36.088  739.981106

In [20]:
new_df.drop(columns = ["continent"])

Country  year  Population  life_exp     gdp_cap
0     Afghanistan  1952     8425333    28.801  779.445314
1     Afghanistan  1957     9240934    30.332  820.853030
2     Afghanistan  1962    10267083    31.997  853.100710
3     Afghanistan  1967    11537966    34.020  836.197138
4     Afghanistan  1972    13079460    36.088  739.981106
...           ...   ...         ...       ...         ...
1699     Zimbabwe  1987     9216418    62.351  706.157306
1700     Zimbabwe  1992    10704340    60.377  693.420786
1701     Zimbabwe  1997    11404948    46.809  792.449960
1702     Zimbabwe  2002    11926563    39.989  672.038623
1703     Zimbabwe  2007    12311143    43.487  469.709298

[1704 rows x 5 columns]

In [44]:
new_df.drop(columns = ["continent"], inplace = True)

new_df.head()

Country  year  Population  life_exp     gdp_cap
0  Afghanistan  1952     8425333    28.801  779.445314
1  Afghanistan  1957     9240934    30.332  820.853030
2  Afghanistan  1962    10267083    31.997  853.100710
3  Afghanistan  1967    11537966    34.020  836.197138
4  Afghanistan  1972    13079460    36.088  739.981106

In [22]:
new_df.drop(columns = ["life_exp", "year"], inplace = True)

new_df.head()

Country  Population     gdp_cap
0  Afghanistan     8425333  779.445314
1  Afghanistan     9240934  820.853030
2  Afghanistan    10267083  853.100710
3  Afghanistan    11537966  836.197138
4  Afghanistan    13079460  739.981106

### <code style="background:yellow;color:black">6. Creating columns:</code>

**In pandas, we can create new columns in a DataFrame using various methods. Here are some common ways to add new columns:**

* **<u><code style="background:yellow;color:black">Using Existing Columns:</u></code>**
    * We can create a new column based on the values of existing columns. 
    * If we want to create a new column from an existing column, we can do like this code below. 
    * <code style="background:yellow;color:black">When we write df["year"] + 7, 7 will be added to every row. So a new column is created and all the values in the new column are year values plus 7.</code> 
    * This is how we can create a new column from a existing column, and there is automatic row wise operation being applied to every new row created in new column.
    * In next code cell below that, we are creating a new column called "gdp". Now, GDP per capita means total gdp of the country divided by its total population.
    * <code style="background:yellow;color:black">So, we are multiplying two existing columns to create a new gdp column.</code> This is how we can create a new column from a existing column, and there is automatic row wise operation being applied to every new row created in new column.<br><br>
    
* **<u><code style="background:yellow;color:black">Creating new columns from scratch:</u></code>**
    * Now, lets say we want to create a new column as a whole, not from any existing column. The syntax is as below.
    * We have to provide the list of all the values that will be present in the new column. As its practically impossible to write 1704 values, so we will write a loop. 
    * We can give list values in various ways, out of which this loop way is one way, but we have to provide all the values/rows of the column.

In [45]:
df.head()

Country  year  Population  life_exp     gdp_cap
0  Afghanistan  1952     8425333    28.801  779.445314
1  Afghanistan  1957     9240934    30.332  820.853030
2  Afghanistan  1962    10267083    31.997  853.100710
3  Afghanistan  1967    11537966    34.020  836.197138
4  Afghanistan  1972    13079460    36.088  739.981106

In [46]:
df["new_col"] = df["year"] + 7

df.head()

Country  year  Population  life_exp     gdp_cap  new_col
0  Afghanistan  1952     8425333    28.801  779.445314     1959
1  Afghanistan  1957     9240934    30.332  820.853030     1964
2  Afghanistan  1962    10267083    31.997  853.100710     1969
3  Afghanistan  1967    11537966    34.020  836.197138     1974
4  Afghanistan  1972    13079460    36.088  739.981106     1979

In [47]:
df["gdp"] = df["gdp_cap"] * df["Population"]

df.head()

Country  year  Population  life_exp     gdp_cap  new_col           gdp
0  Afghanistan  1952     8425333    28.801  779.445314     1959  6.567086e+09
1  Afghanistan  1957     9240934    30.332  820.853030     1964  7.585449e+09
2  Afghanistan  1962    10267083    31.997  853.100710     1969  8.758856e+09
3  Afghanistan  1967    11537966    34.020  836.197138     1974  9.648014e+09
4  Afghanistan  1972    13079460    36.088  739.981106     1979  9.678553e+09

In [48]:
df["new"] = [ i + 3 for i in range(1704)]

df

Country  year  Population  life_exp     gdp_cap  new_col  \
0     Afghanistan  1952     8425333    28.801  779.445314     1959   
1     Afghanistan  1957     9240934    30.332  820.853030     1964   
2     Afghanistan  1962    10267083    31.997  853.100710     1969   
3     Afghanistan  1967    11537966    34.020  836.197138     1974   
4     Afghanistan  1972    13079460    36.088  739.981106     1979   
...           ...   ...         ...       ...         ...      ...   
1699     Zimbabwe  1987     9216418    62.351  706.157306     1994   
1700     Zimbabwe  1992    10704340    60.377  693.420786     1999   
1701     Zimbabwe  1997    11404948    46.809  792.449960     2004   
1702     Zimbabwe  2002    11926563    39.989  672.038623     2009   
1703     Zimbabwe  2007    12311143    43.487  469.709298     2014   

               gdp   new  
0     6.567086e+09     3  
1     7.585449e+09     4  
2     8.758856e+09     5  
3     9.648014e+09     6  
4     9.678553e+09     7  
...            ...   ...  
1699  6.508241e+09  1702  
1700  7.422612e+09  1703  
1701  9.037851e+09  1704  
1702  8.015111e+09  1705  
1703  5.782658e+09  1706  

[1704 rows x 8 columns]

**We are removing newly created columns, which we created just to show column creation. We dont need them anymore.**

In [49]:
df.drop(columns = ["new_col", "gdp", "new"], inplace = True)

df.head()

Country  year  Population  life_exp     gdp_cap
0  Afghanistan  1952     8425333    28.801  779.445314
1  Afghanistan  1957     9240934    30.332  820.853030
2  Afghanistan  1962    10267083    31.997  853.100710
3  Afghanistan  1967    11537966    34.020  836.197138
4  Afghanistan  1972    13079460    36.088  739.981106

<hr style="border: 1px solid gray;">

# <code style="background:yellow;color:black">Dataframe indexing:</code>

**<code style="background:yellow;color:black">DataFrame indexing</code> in pandas refers to the process of selecting specific rows and/or columns from a DataFrame. Indexing allows us to retrieve, filter, and manipulate data in a DataFrame based on certain criteria.** 

### <code style="background:yellow;color:black">Implicit and Explicit indexes:</code>

**There are two types of indexes in pandas dataframe: <code style="background:yellow;color:black">Implicit and Explicit.</code> In pandas, the terms "implicit index" and "explicit index" refer to different ways of labeling and referencing data within a DataFrame or Series.**

**<u><code style="background:yellow;color:black">Implicit index:</code></u>** 

* **An implicit index is the default integer-based index that is automatically assigned to each row in a DataFrame or element in a Series when the data structure is created.**
* Implicit index is same as python index. Python indexing starts from 0 right, so does implicit index.
* **<code style="background:yellow;color:black">Implicit index always starts from 0.</code> So, in backend it is always index 0,1,2,...**
* The implicit index is created automatically unless a custom index is specified.

**<u><code style="background:yellow;color:black">Explicit index:</code></u>** 

* **An explicit index is a user-defined set of labels assigned to each row or element.**
* Whatever we see here on screen is called **<code style="background:yellow;color:black">explicit index.</code>** 
* Explicit index can be any immutable datatype, such as integers, strings, or datetime objects.
* **An explicit index is specified by the user when creating the DataFrame or Series or by modifying the index later using methods like set_index() method.**

### <code style="background:yellow;color:black">How to access these indices or in other words how to access various rows:</code>

* Pandas have provided us with different ways to access implicit and explicit indices or rows of a DataFrame. 
* **There is an indexing called iloc which will be used to access implicit indices and whenever we want to access explicit indices, we use an indexing called loc**. 
* Easiest way to remember is - iloc word starts with i and implicit word starts with i as well. 
* **<code style="background:yellow;color:black">Implicit index --> 0 --> iloc</code>**
* **<code style="background:yellow;color:black">Explicit index --> whatever is visible --> loc</code>**

Now lets understand the use cases of these two.


There are various ways to perform indexing in pandas, and here are some common methods:**

* **df.columns gives us all the column names.**
* **df.index.values gives us a numpy array of all the row indices i.e. number of rows.**

In [193]:
df.head()

Country  year  Population  life_exp     gdp_cap
0  Afghanistan  1952     8425333    28.801  779.445314
1  Afghanistan  1957     9240934    30.332  820.853030
2  Afghanistan  1962    10267083    31.997  853.100710
3  Afghanistan  1967    11537966    34.020  836.197138
4  Afghanistan  1972    13079460    36.088  739.981106

In [194]:
df.columns

Index(['Country', 'year', 'Population', 'life_exp', 'gdp_cap'], dtype='object')

In [195]:
df.index.values

array([   0,    1,    2, ..., 1701, 1702, 1703], dtype=int64)

**The row indices 0 to 1703 can be anything, it dont necessarily have to serial wise numbers.** We can change/update them. Below is the way of updating or changing row indices. We have to provide all the new indices's values as a list. We can see now index starts from 3 instead of 0 and all the way upto 1706.

**Note that here we are assigning row indices, so no need to write inplace = True.** 

"inplace = True" we generally write during calling functions like drop(). 

In [196]:
df.index = [ i + 3 for i in range(1704)]

In [197]:
df.head()

Country  year  Population  life_exp     gdp_cap
3  Afghanistan  1952     8425333    28.801  779.445314
4  Afghanistan  1957     9240934    30.332  820.853030
5  Afghanistan  1962    10267083    31.997  853.100710
6  Afghanistan  1967    11537966    34.020  836.197138
7  Afghanistan  1972    13079460    36.088  739.981106

What is the index that is visible to us for the first row? - 3. 

Now 3 is explicit because it is what is visible on screen. Now if we do df.loc[3] we are trying to access explicit index 3. It returns first complete row as an output. 

In [199]:
df.loc[3]

Country       Afghanistan
year                 1952
Population        8425333
life_exp           28.801
gdp_cap        779.445314
Name: 3, dtype: object

In [200]:
df.loc[5]

Country       Afghanistan
year                 1962
Population       10267083
life_exp           31.997
gdp_cap         853.10071
Name: 5, dtype: object

Now, from python perspective, what is the index of first row? - 0. 

We use iloc to access implicit index. So if we do df.iloc[0] we are trying to access implicit index 0. It returns first complete row as an output. 

So df.loc[3] and df.iloc[0] are both returning the first row as their output.

In [201]:
df.iloc[0]

Country       Afghanistan
year                 1952
Population        8425333
life_exp           28.801
gdp_cap        779.445314
Name: 3, dtype: object

In [202]:
df.head()

Country  year  Population  life_exp     gdp_cap
3  Afghanistan  1952     8425333    28.801  779.445314
4  Afghanistan  1957     9240934    30.332  820.853030
5  Afghanistan  1962    10267083    31.997  853.100710
6  Afghanistan  1967    11537966    34.020  836.197138
7  Afghanistan  1972    13079460    36.088  739.981106

The major difference between loc and iloc is this. Remember the python lists slicing concept. In slicing for example, 1:4, the end index i.e. 4 is not included in the slice that we make. This is where loc and iloc differs. 

**In loc, if we write df.loc[3:6], end index is inclusive**. We get rows with explicit indices 3,4,5,6 as output. 

**In iloc, if we write df.iloc[0:3], end is not inclusive**. We get rows with implicit indexes 0,1,2 only, and **not 3** which is the end index. This is very similar to python. Remember index 3 which is visible on screen is index 0 internally. 

In [203]:
df.loc[3:6] 

Country  year  Population  life_exp     gdp_cap
3  Afghanistan  1952     8425333    28.801  779.445314
4  Afghanistan  1957     9240934    30.332  820.853030
5  Afghanistan  1962    10267083    31.997  853.100710
6  Afghanistan  1967    11537966    34.020  836.197138

In [204]:
df.iloc[0:3]

Country  year  Population  life_exp     gdp_cap
3  Afghanistan  1952     8425333    28.801  779.445314
4  Afghanistan  1957     9240934    30.332  820.853030
5  Afghanistan  1962    10267083    31.997  853.100710

In [205]:
df.head(10)

Country  year  Population  life_exp     gdp_cap
3   Afghanistan  1952     8425333    28.801  779.445314
4   Afghanistan  1957     9240934    30.332  820.853030
5   Afghanistan  1962    10267083    31.997  853.100710
6   Afghanistan  1967    11537966    34.020  836.197138
7   Afghanistan  1972    13079460    36.088  739.981106
8   Afghanistan  1977    14880372    38.438  786.113360
9   Afghanistan  1982    12881816    39.854  978.011439
10  Afghanistan  1987    13867957    40.822  852.395945
11  Afghanistan  1992    16317921    41.674  649.341395
12  Afghanistan  1997    22227415    41.763  635.341351

In [206]:
df.loc[5:8]

Country  year  Population  life_exp     gdp_cap
5  Afghanistan  1962    10267083    31.997  853.100710
6  Afghanistan  1967    11537966    34.020  836.197138
7  Afghanistan  1972    13079460    36.088  739.981106
8  Afghanistan  1977    14880372    38.438  786.113360

In [207]:
df.iloc[5:8]

Country  year  Population  life_exp     gdp_cap
8   Afghanistan  1977    14880372    38.438  786.113360
9   Afghanistan  1982    12881816    39.854  978.011439
10  Afghanistan  1987    13867957    40.822  852.395945

In [208]:
df.head()

Country  year  Population  life_exp     gdp_cap
3  Afghanistan  1952     8425333    28.801  779.445314
4  Afghanistan  1957     9240934    30.332  820.853030
5  Afghanistan  1962    10267083    31.997  853.100710
6  Afghanistan  1967    11537966    34.020  836.197138
7  Afghanistan  1972    13079460    36.088  739.981106

Now, we know multi indexing as well from python lists. 

So, this df.iloc[[2,4]] will return implicit index 2th row and implicit index 4th row, which is visible row 5 and 7 respectively.

In [209]:
df.iloc[[2,4]]

Country  year  Population  life_exp     gdp_cap
5  Afghanistan  1962    10267083    31.997  853.100710
7  Afghanistan  1972    13079460    36.088  739.981106

In [210]:
# In python, -1 index stands for last element. so, negative index exists here as well.

df.iloc[-1]

Country         Zimbabwe
year                2007
Population      12311143
life_exp          43.487
gdp_cap       469.709298
Name: 1706, dtype: object

In [211]:
# df.tail() just to check last row

df.tail()

Country  year  Population  life_exp     gdp_cap
1702  Zimbabwe  1987     9216418    62.351  706.157306
1703  Zimbabwe  1992    10704340    60.377  693.420786
1704  Zimbabwe  1997    11404948    46.809  792.449960
1705  Zimbabwe  2002    11926563    39.989  672.038623
1706  Zimbabwe  2007    12311143    43.487  469.709298

**There is no explicit index called -1, so the following code will return error**. Negetive indexing will work for iloc, negetive indexing might not work for loc, because if you dont have explicit negetive index, if your negetive indexes are explicit then you can access using that. So, this is more like dict key value type if -1 key doesnt exist it will throw keyerror.

In [212]:
df.loc[-1]

KeyError: -1

In [213]:
df.head()

Country  year  Population  life_exp     gdp_cap
3  Afghanistan  1952     8425333    28.801  779.445314
4  Afghanistan  1957     9240934    30.332  820.853030
5  Afghanistan  1962    10267083    31.997  853.100710
6  Afghanistan  1967    11537966    34.020  836.197138
7  Afghanistan  1972    13079460    36.088  739.981106

<hr style="border: 1px solid gray;">

# <code style="background:yellow;color:black">set_index() method:</code>

* **In pandas, the <code style="background:yellow;color:black">set_index() method</code> is used to set a specific column or a combination of columns as the DataFrame's index. The index of a DataFrame is like its "row labels" and helps in identifying and accessing rows more efficiently.**
* Setting an index can be beneficial for various operations, such as merging/joining DataFrames, grouping data, and performing certain types of data analysis.
* **With this method, we are basically setting the index to some other value.**
* **What this below code cell <code style="background:yellow;color:black">df.set_index("Country")</code> does is like--** I am saying consider country column as my index and store in a copied dataframe called temp. In the output of this it is clearly seen that row indices are changed to country names. **So row indices can be anything.** 
* **How to access Afghanistan row indices? Like the code below.** Because the row indices are visible on screen means they are explicit, we use loc. It will give us all the rows where its afghanistan. Again these are explicit indices, because implicit are always 0,1,2,3,...

In [9]:
temp = df.set_index("Country")

temp.head()

year  Population continent  life_exp     gdp_cap
Country                                                      
Afghanistan  1952     8425333      Asia    28.801  779.445314
Afghanistan  1957     9240934      Asia    30.332  820.853030
Afghanistan  1962    10267083      Asia    31.997  853.100710
Afghanistan  1967    11537966      Asia    34.020  836.197138
Afghanistan  1972    13079460      Asia    36.088  739.981106

In [216]:
temp.loc["Afghanistan"]

year  Population  life_exp     gdp_cap
Country                                            
Afghanistan  1952     8425333    28.801  779.445314
Afghanistan  1957     9240934    30.332  820.853030
Afghanistan  1962    10267083    31.997  853.100710
Afghanistan  1967    11537966    34.020  836.197138
Afghanistan  1972    13079460    36.088  739.981106
Afghanistan  1977    14880372    38.438  786.113360
Afghanistan  1982    12881816    39.854  978.011439
Afghanistan  1987    13867957    40.822  852.395945
Afghanistan  1992    16317921    41.674  649.341395
Afghanistan  1997    22227415    41.763  635.341351
Afghanistan  2002    25268405    42.129  726.734055
Afghanistan  2007    31889923    43.828  974.580338

In [217]:
# df remains unchanged because we have changed indexes of temp dataframe.

df.head()

Country  year  Population  life_exp     gdp_cap
3  Afghanistan  1952     8425333    28.801  779.445314
4  Afghanistan  1957     9240934    30.332  820.853030
5  Afghanistan  1962    10267083    31.997  853.100710
6  Afghanistan  1967    11537966    34.020  836.197138
7  Afghanistan  1972    13079460    36.088  739.981106

<hr style="border: 1px solid gray;">

# <code style="background:yellow;color:black">reset_index() method:</code>

* **The <code style="background:yellow;color:black">reset_index() method</code> in pandas is used to reset the index of a DataFrame. When we set an index using the set_index() method or similar operations, the existing index is replaced with the specified column(s), and the DataFrame may no longer have a default integer-based index. The reset_index() method allows us to revert to the default integer-based index and move the current index back to regular columns.**
* With this methos we can reset the index back to original.**
* **In the below code cell, reset_index() method is used.** Notice that in the output of the code below, 0,1,2,3...1703 is resetted. And, original index 3,4,5,6,...1706 is created as a new column.
* **If we dont want that new index column of 3,4,5,6,...1706, we can implement the code below. What it does is it automatically resets and removes that index column as well. So it resets and removes unnecessary column.**
* Now, in code below to below cell "df.reset_index(drop = True)", it showed us the required output, but the code df.head() doesnt show us that required output. 
* Why? - Because we havent made it permanent by saying inplace = True. So now if we check df.head(), we get back the original dataset.

In [218]:
df.reset_index()

index      Country  year  Population  life_exp     gdp_cap
0         3  Afghanistan  1952     8425333    28.801  779.445314
1         4  Afghanistan  1957     9240934    30.332  820.853030
2         5  Afghanistan  1962    10267083    31.997  853.100710
3         6  Afghanistan  1967    11537966    34.020  836.197138
4         7  Afghanistan  1972    13079460    36.088  739.981106
...     ...          ...   ...         ...       ...         ...
1699   1702     Zimbabwe  1987     9216418    62.351  706.157306
1700   1703     Zimbabwe  1992    10704340    60.377  693.420786
1701   1704     Zimbabwe  1997    11404948    46.809  792.449960
1702   1705     Zimbabwe  2002    11926563    39.989  672.038623
1703   1706     Zimbabwe  2007    12311143    43.487  469.709298

[1704 rows x 6 columns]

In [219]:
df.reset_index(drop = True)

Country  year  Population  life_exp     gdp_cap
0     Afghanistan  1952     8425333    28.801  779.445314
1     Afghanistan  1957     9240934    30.332  820.853030
2     Afghanistan  1962    10267083    31.997  853.100710
3     Afghanistan  1967    11537966    34.020  836.197138
4     Afghanistan  1972    13079460    36.088  739.981106
...           ...   ...         ...       ...         ...
1699     Zimbabwe  1987     9216418    62.351  706.157306
1700     Zimbabwe  1992    10704340    60.377  693.420786
1701     Zimbabwe  1997    11404948    46.809  792.449960
1702     Zimbabwe  2002    11926563    39.989  672.038623
1703     Zimbabwe  2007    12311143    43.487  469.709298

[1704 rows x 5 columns]

In [220]:
df.head()

Country  year  Population  life_exp     gdp_cap
3  Afghanistan  1952     8425333    28.801  779.445314
4  Afghanistan  1957     9240934    30.332  820.853030
5  Afghanistan  1962    10267083    31.997  853.100710
6  Afghanistan  1967    11537966    34.020  836.197138
7  Afghanistan  1972    13079460    36.088  739.981106

In [10]:
df.reset_index(drop = True, inplace = True)

df.head()

Country  year  Population continent  life_exp     gdp_cap
0  Afghanistan  1952     8425333      Asia    28.801  779.445314
1  Afghanistan  1957     9240934      Asia    30.332  820.853030
2  Afghanistan  1962    10267083      Asia    31.997  853.100710
3  Afghanistan  1967    11537966      Asia    34.020  836.197138
4  Afghanistan  1972    13079460      Asia    36.088  739.981106

<hr style="border: 1px solid gray;">

### <code style="background:yellow;color:black">Difference between pandas method and function:</code> 

* **<code style="background:yellow;color:black">When we have created a dataframe object "df" (or any other variable name), any function which is applied on top of this pandas dataframe object is a method.</code>** 
* df refers to a specific dataframe object on which we are applying a method. 
* These are typically methods that are called on an existing dataFrame object for example df.mean(), df.drop(), df.groupby(). 
* These methods operate directly on the dataFrame, and they can access and modify its data and structure.

<br>

* **<code style="background:yellow;color:black">When we say pd.any_function_name(), it refers to a function that has been written inside pandas module.</code>** 
* pd refers to the pandas library itself. 
* These are functions provided by the pandas library, which we call with dataFrame or series objects as arguments for example pd.read_csv(), pd.concat(), pd.merge(). 
* These functions usually take dataframe or series objects as input and return new objects or perform operations that don't modify the original dataFrame directly.